p2p/server.go

```diff
@@ -289,50 +338,161 @@ func (srv *Server) startListening() error {
+			}
+		}
+
+		select {
+		case <-srv.quit:
+			// The server was stopped. Run the cleanup logic.
+			glog.V(logger.Detail).Infoln("<-quit: spinning down")
+			break running
+		case n := <-srv.addstatic:
+			// This channel is used by AddPeer to add to the
+			// ephemeral static peer list. Add it to the dialer,
+			// it will keep the node connected.
+			glog.V(logger.Detail).Infoln("<-addstatic:", n)
+			dialstate.addStatic(n)
+		case op := <-srv.peerop:
+			// This channel is used by Peers and PeerCount.
+			op(peers)
+			srv.peeropResp <- struct{}{}
+		case t := <-taskdone:
+			// A task got done. Tell the dialer about it so it
+			// can update its state and remove it from the active
+			// tasks list.
+			glog.V(logger.Detail).Infoln("<-taskdone:", t)
+			dialstate.taskDone(t, now)
+			deltask(t)
+		case c := <-srv.posthandshake:
+			// A connection has passed the encryption handshake so
+			// the remote identity is known (but hasn't been verified yet).
+			if trusted[c.id] {
+				// Ensure that the trusted flag is set before checking against MaxPeers.
+				c.flags |= trustedConn
+			}
+			glog.V(logger.Detail).Infoln("<-posthandshake:", c)
+			// TODO: track in-progress inbound node IDs (pre-Peer) to avoid dialing them.
+			c.cont <- srv.encHandshakeChecks(peers, c)
+		case c := <-srv.addpeer:
+			// At this point the connection is past the protocol handshake.
+			// Its capabilities are known and the remote identity is verified.
+			glog.V(logger.Detail).Infoln("<-addpeer:", c)
+			err := srv.protoHandshakeChecks(peers, c)
+			if err != nil {
+				glog.V(logger.Detail).Infof("Not adding %v as peer: %v", c, err)
+			} else {
+				// The handshakes are done and it passed all checks.
+				p := newPeer(c, srv.Protocols)
+				peers[c.id] = p
+				go srv.runPeer(p)
+			}
+			// The dialer logic relies on the assumption that
+			// dial tasks complete after the peer has been added or
+			// discarded. That's why the task get unblocked last.
+			c.cont <- err
+		case p := <-srv.delpeer:
+			// A peer disconnected.
+			glog.V(logger.Detail).Infoln("<-delpeer:", p)
+			delete(peers, p.ID())
+		}
 	}
-	srv.running = false
-	srv.lock.Unlock()
 
-	glog.V(logger.Info).Infoln("Stopping Server")
+	// Terminate discovery. If there is a running lookup it will terminate soon.
 	srv.ntab.Close()
-	if srv.listener != nil {
-		// this unblocks listener Accept
-		srv.listener.Close()
+	// Disconnect all peers.
+	for _, p := range peers {
+		p.Disconnect(DiscQuitting)
+	}
+	// Wait for peers to shut down. Pending connections are
+	// not handled here and will terminate soon-ish because srv.quit
+	// is closed.
+	// TODO: tracking post-handshake conns could help here.
+	for len(peers) > 0 {
+		select {
+		case t := <-taskdone:
+			glog.V(logger.Detail).Infoln("<-taskdone (spindown):", t)
+			deltask(t)
+		case p := <-srv.delpeer:
+			glog.V(logger.Detail).Infoln("<-delpeer (spindown):", p)
+			delete(peers, p.ID())
+		}
```

Comment:
If you were to add ` || len(tasks) > 0` in there too then I can imagine it's ok (and actually nice not to leak the task goroutines). Of course, that could mean that you need to possibly wait 30s for the dial expiration task.

Another solution would be to drop the task done depletion here, and add a select into task termination to check for server quit.
```
select {
   case taskdone <- t:
   case <-srv.quit:
}
```